python/lsst/ts/pmd/component.py

```diff
@@ -229,27 +212,35 @@ async def multiplexer_recovery(self):
         if not self.connected:
             raise RuntimeError("Not connected")
 
-        # enter the config interface
-        reply = await self.send_msg("SPC")
-        await asyncio.sleep(5)
-        self.log.debug(f"Multiplexer SPC response is: {reply}")
-        reply3 = self.send_msg("QU")
-        await asyncio.sleep(2)
-        self.log.debug(f"{reply3=}")
-
-        # now check the sensors
-        positions = await self.get_slots_position()
-        success = True
-        for i, name in enumerate(self.names):
-            # Skip the channels that have nothing configured
-            if name == "":
-                continue
-            elif positions[i]:
-                self.log.debug(f"Sensor {i} = {positions[i]:0.3f}")
-            else:
-                self.log.error(
-                    f"Sensor {i} returned a blank string. " "Multiplexer reset failed."
-                )
-                success = False
+        read_error_count = 1
```

Comment:
Now you have mulitplexer_recovery calling get_slots_position and get_slots_position calling multiplexer_recovery with no sort of guard to avoid an infinite loop. This is very confusing and potentially a big problem. Please try to clean up the logic.

I am not sure I fully understand the logic you are aiming for. Please add a code comment to get_slots_position that explains it, for future readers. If that logic is as as follows:
* Try to read all configured channels
* If any channels return a plain "\r" then try recovering the the mutiplexor and then read again
* If 3 attempts at recovery fail, then give up and return nan for the positions that could not be read

then I strongly feel it is better NOT to have multiplexer_recovery try to read positions (like your first try). I'd be inclined to do it something like this (I have taken the liberty of renaming some methods and variables to possibly improve readability):

```
def basic_get_positions(self):
    """Read the position from each channel with a non-empty name.

    Do not retry on error.

    Returns
    -------
    * positions: a list of positions, one per entry in self.name,
      with NaN for each channel with a blank name or that could not be read.
    * isok: if True then all named channels were read.
    """
    if not self.connected:
        raise Exception("Not connected")
    positions = [math.nan] * len(self.names)
    for i, name in enumerate(self.names):
        # Skip the channels that have nothing configured
        if name == "":
            continue
        reply = await self.send_msg(str(i + 1))
        if reply != b"\r":
            split_reply = reply.decode().split(":")
            positions[i] = float(split_reply[-1])
            self.log.debug(f"positions[{i}]={positions}")
        else:
            return positions, False
    return positions, True


def get_positions(self, max_resets=3):
    """Read the position from each channel with a non-empty name.

    Try to reset the multiplexor up to max_resets times,
    (and read positions again) if any configured channel cannot be read.

    Returns
    -------
    * positions: a list of positions, one per entry in self.name,
      with NaN for each channel with a blank name or that could not be read.
    * isok: if True then all named channels were read.
    """
    if max_resets < 0:
        raise ValueError(f"{max_resets=} must be >= 0")
    for ntries in range(max_resets+1):
       positions, isok = await self.basic_read_positions()
       if isok:
            break
       await self.reset_mutiplexor()
    return positions, isok


def reset_mutiplexor(self):
    """Try to reset the multiplexor."""
    reply = await self.send_msg("SPC")
    await asyncio.sleep(5)
    self.log.debug(f"Multiplexer SPC response is: {reply}")
    reply = self.send_msg("QU")
    await asyncio.sleep(2)
    self.log.debug(f"Multiplexer QU response is: {reply}")
```

Reply:
BTW: I made max_resets a method argument, but it could just as easily be a module constant or a configuration parameter (well worth considering).